crates/concrete_codegen_mlir/src/codegen.rs

```diff
@@ -1,1004 +1,962 @@
-            TypeSpec::Simple { name, .. } => signed.contains(&name.name.as_str()),
-            TypeSpec::Generic { name, .. } => signed.contains(&name.name.as_str()),
-            TypeSpec::Array { .. } => unreachable!(),
-        }
-    }
+    Ok(())
+}
 
-    fn is_float(&self, type_info: &TypeSpec) -> bool {
-        let signed = ["f32", "f64"];
-        match type_info {
-            TypeSpec::Simple { name, .. } => signed.contains(&name.name.as_str()),
-            TypeSpec::Generic { name, .. } => signed.contains(&name.name.as_str()),
-            TypeSpec::Array { .. } => unreachable!(),
-        }
-    }
+#[derive(Debug, Clone, Copy)]
+struct FunctionCodegenCtx<'a> {
+    pub module_ctx: ModuleCodegenCtx<'a>,
+    pub function_id: DefId,
 }
 
-fn compile_module(
-    session: &Session,
-    context: &MeliorContext,
-    mlir_module: &MeliorModule,
-    ast_helper: &AstHelper<'_>,
-    module_info: &ModuleInfo<'_>,
-    module: &Module,
-) -> Result<(), Box<dyn Error>> {
-    let body = mlir_module.body();
-
-    let mut imports = HashMap::new();
-
-    for import in &module.imports {
-        let target_module = ast_helper
-            .get_module_from_import(&import.module)
-            .unwrap_or_else(|| {
-                panic!(
-                    "failed to find import {:?} in module {}",
-                    import, module.name.name
-                )
-            });
-
-        for symbol in &import.symbols {
-            imports.insert(symbol.name.clone(), target_module);
-        }
+impl<'a> FunctionCodegenCtx<'a> {
+    pub fn get_fn_body(&self) -> &FnBody {
+        self.module_ctx
+            .ctx
+            .program
+            .functions
+            .get(&self.function_id)
+            .expect("function body should exist")
     }
 
-    let scope_ctx: ScopeContext = ScopeContext {
-        locals: Default::default(),
-        function: None,
-        module_info,
-        imports,
-    };
-
-    for statement in &module.contents {
-        match statement {
-            ModuleDefItem::Constant(_) => todo!(),
-            ModuleDefItem::Function(info) => {
-                let mut scope_ctx = scope_ctx.clone();
-                scope_ctx.function = Some(info.clone());
-                let op = compile_function_def(session, context, &scope_ctx, info)?;
-                body.append_operation(op);
-            }
-            ModuleDefItem::Struct(_) => todo!(),
-            ModuleDefItem::Type(_) => todo!(),
-            ModuleDefItem::Module(info) => {
-                let module_info = module_info.modules.get(&info.name.name).unwrap_or_else(|| {
-                    panic!(
-                        "submodule {} not found while compiling module {}",
-                        info.name.name, module.name.name
-                    )
-                });
-                compile_module(session, context, mlir_module, ast_helper, module_info, info)?;
-            }
-        }
+    pub fn get_fn_sig(&self) -> &(Vec<Ty>, Ty) {
```

Comment:
since this function is a public, can you rename it to `get_fn_signature`  for more clarity?